### PR TITLE
CAMS-55 configuration keyvault

### PR DESCRIPTION
**Pull request by magile32**
# Purpose

Prepare Azure Keyvault for storing application secrets and configuration.

# Major Changes

- Bicep deployment script to provision Azure Keyvault and Managed Identity with vault access
- Bicep deployment script to configure private endpoint and dns zone to enable and secure Azure Keyvault access
- Updating existing backend deployment to set user identity for keyvault access
- Edited az-func-deploy.sh script to generate KeyVault References for Values Stored in KeyVault

# Testing/Validation

- Validate creation and configuration of Azure resources related to Keyvault
- Verify existing deployment enables backend to access secrets in vault
- Deleted and recreated environment along with kv resources to verify proper resource creation, configuration and app functionality

# Notes

The following Github Action secrets added:
```
AZ_KV_APP_CONFIGURATION [key vault name]
AZ_KV_APP_CONFIGURATION_ID [managed identity with key vault access]
...
```

Deployment script for backend was updated to include two new options `--kvName` and `--kvSettings`. kvName refers to an existing keyvault storing application secrets. kvSettings is the space separated list of env key names expected to be used by the application. This additional will setup the Azure Keyvault references in functionapp App Settings.

### CODE REVIEW

File: ops/cloud-deployment/ustp-cams-kv-app-config-setup.bicep [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+param privateEndpointSubnetName string
+
+@description('Subnet prefix')
+param privateEndpointSubnetPrefix string
+
+@description('Resource group the subnet resides within')
+param networkResourceGroup string
+
+module kvAppConfig './keyvault-app-config-setup/kv-app-config.bicep' = {
+  name: '${stackName}-kv-app-config-module'
```
Comment (magile32) [inline ops/cloud-deployment/ustp-cams-kv-app-config-setup.bicep:28]: Let's rename this

File: .github/workflows/continuous-deployment.yml [hunk 2]
```diff
@@ -32,7 +32,7 @@ on:
 jobs:
   build-info:
     runs-on: ubuntu-latest
-    environment: ${{ endsWith(github.ref,'refs/heads/main') && 'Main-Gov' || 'Develop' }}
+    environment: ${{ endsWith(github.ref,'refs/heads/main') && 'Main-Gov' || 'Test-KV-Config' }}
```
Comment (magile32) [inline .github/workflows/continuous-deployment.yml:35]:
```suggestion
    environment: ${{ endsWith(github.ref,'refs/heads/main') && 'Main-Gov' || 'Develop }}
```

Comment (magile32) [reply .github/workflows/continuous-deployment.yml:35]: Will need to revert this back to `Develop` and clean up Test-KV-Config environment in GHA.

File: .github/workflows/continuous-deployment.yml [hunk 3]
```diff
@@ -469,18 +479,12 @@ jobs:
...
-          MSSQL_DATABASE=${{ secrets.AZ_MSSQL_DATABASE }} \
-          MSSQL_DATABASE_DXTR=${{ secrets.AZ_MSSQL_DATABASE_DXTR }} \
-          MSSQL_USER=${{ secrets.AZ_MSSQL_USER }} \
-          MSSQL_PASS=${{ secrets.AZ_MSSQL_PASS }} \
-          MSSQL_ENCRYPT=${{ secrets.AZ_MSSQL_ENCRYPT }} \
-          MSSQL_TRUST_UNSIGNED_CERT=${{ secrets.AZ_MSSQL_TRUST_UNSIGNED_CERT }} \
-          STARTING_MONTH=${{ vars.STARTING_MONTH }} \
+          --kvName ${{ secrets.AZ_KV_APP_CONFIG_NAME }} \
+          --kvSettings "MSSQL_HOST MSSQL_DATABASE_DXTR MSSQL_USER MSSQL_PASS MSSQL_ENCRYPT MSSQL_TRUST_UNSIGNED_CERT FEATURE_FLAG_SDK_KEY" \
+          --settings "STARTING_MONTH=${{ vars.STARTING_MONTH }} \
```
Comment (magile32) [inline .github/workflows/continuous-deployment.yml:484]: Do we want to move `STARTING_MONTH` into keyvault?

File: docs/deployment.md [hunk 4]
```diff
@@ -2,67 +2,70 @@
 
 ## Pipeline Environment Variables
 
-Note required environment variables and secrets defined in build tool for pipeline execution in Flexion and  **shared** with USTP.
+Note required environment variables and secrets defined in build tool for pipeline execution in Flexion and **shared** with USTP.
```
Comment (magile32) [inline docs/deployment.md:5]: What markdown linter was used here? Should we add to pre-commit if possible.

File: ops/cloud-deployment/params/ustp-cams-kv-appconfig.parameters.json
Comment (magile32) [inline ops/cloud-deployment/params/ustp-cams-kv-appconfig.parameters.json:1]: We don't have to check in the file. This was passed around for testing purposes.

**Comment by governmentSponsored** [comment]
When do we plan on testing the deployment of this in USTP env?

File: .github/workflows/continuous-deployment.yml (hunk 3 above)
Comment (governmentSponsored) [reply .github/workflows/continuous-deployment.yml:484]: Nope, makes sense in LD or as a normal env var

File: ops/cloud-deployment/params/ustp-cams-kv-appconfig.parameters.json
Comment (amorrow-flexion) [inline ops/cloud-deployment/params/ustp-cams-kv-appconfig.parameters.json:1]: Add this to the .gitignore

**Review by governmentSponsored — APPROVED** [review]
worked deploying to USTP env

**Pull request merged**